nltk/collocations.py

```diff
@@ -176,12 +176,14 @@ def score_ngram(self, score_fn, w1, w2):
         function.  Following Church and Hanks (1990), counts are scaled by
         a factor of 1/(window_size - 1).
         """
-        n_all = self.N
-        n_ii = self.ngram_fd[(w1, w2)] / (self.window_size - 1.0)
+        n_all = self.ngram_fd.N()
+        n_ii = self.ngram_fd[(w1, w2)]/(self.window_size - 1.0)
         if not n_ii:
             return
-        n_ix = self.word_fd[w1]
-        n_xi = self.word_fd[w2]
+        nix_list = [tup for tup in self.ngram_fd if tup[0] == w1]
```

Comment:
Stylistically, I'd much prefer to see the following than introduce another var:
```python
        n_ix = sum(freq for (ng1, ng2), freq in iteritems(self.ngram_fd)
                   if ng == w1)
```

However, this code is irrelevant. The approach adds unjustifiable computational complexity. The incumbent approach is an approximation, one suggested by textbooks, etc., for efficiency; and yes there appear to be cases where it breaks.

Since a common use-case applies this function for a large quantity of n-grams, we need to get `n_ix` for a particular n-gram in O(1), or maybe at worst O(number of contexts for w1), but the proposed approach uses O(vocabulary) which is too much. So if we go down this route numerically, we need to pre-store `n_ix` and `n_xi`; or my preferred solution is to actually store the number (where greater than 0) that one should subtract from `word_fd[w1]` (`[w2]`) to get `n_ix` (`n_xi`); the data structure will be sparsely populated.